### PR TITLE
Drop-in file systemd configuration

### LINKED ISSUE #1009: Add a way to set the open files limit in systemd service
Hello,

The sysconfig file contains a variable (MAXOPENFILES) that is used to set the limit for the opened files.

The problem is that under RHEL7, the hard limit set by systemd is 4096, that means that nothing can change that value to an higher value.

The .service file should be modified to allow setting this to an higher value (with probably the same default as the one in sysconfig, 8192)

### PR DESCRIPTION
**Pull request by jan-win1993**
#### Pull Request (PR) description

Currently the Systemd drop-in file can only override environment variables.
I would like to expand this functionality to make it possible to override every configuration in the **[Service]** section of the drop-in file.

#### This Pull Request (PR) fixes the following issues

Fixes #1009

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -281,6 +284,7 @@
   Boolean $manage_service                         = true,
   Boolean $service_enable                         = true,
   Enum['running', 'stopped'] $service_ensure      = 'running',
+  Optional[Hash] $service_override                = undef,
```
Comment (ekohl) [inline manifests/init.pp:287]: I'd simplify this to:
```suggestion
  Hash $service_override                = {},
```

You'll need to modify all other instances as well, but this simplifies things. Also, the type in the template itself is stricter than what you have here. I'd make this the same so puppet-strings can properly extract it.

File: templates/jenkins-override.epp [hunk 2]
```diff
@@ -1,5 +1,12 @@
-<%| Hash[String, Struct[{value => Any}]] $environment |-%>
+<%- | Hash[String, Struct[{value => Any}]] $environment,
+      Optional[Hash[String, Struct[{value => Any}]]] $dropin_config = undef
```
Comment (ekohl) [inline templates/jenkins-override.epp:2]: Do you expect to place entries here? Or why do you have the struct as a value with just a single entry (`value`)?

I think for the env vars I kept it to the complex type (with `value`) out of backwards compatibility for the parameter but for a new parameter I'd consider making it simple.

**Review by ekohl — COMMENTED** [review]
It looks like all tests fail. I'm not sure if that's related.

**Comment by ekohl** [comment]
Please rebase. I just merged https://github.com/voxpupuli/puppet-jenkins/pull/1041 which means the tests should be green.

File: templates/jenkins-override.epp [hunk 3]
```diff
@@ -1,5 +1,12 @@
-<%| Hash[String, Struct[{value => Any}]] $environment |-%>
+<%- | Hash[String, Struct[{value => Any}]] $environment,
+      Optional[Hash[String, String]] $dropin_config = undef
```
Comment (ekohl) [inline templates/jenkins-override.epp:2]: I'd also make this a Hash
```suggestion
      Hash[String, String] $dropin_config = {},
```

File: templates/jenkins-override.epp [hunk 4]
```diff
@@ -1,5 +1,12 @@
...
+| -%>
 [Service]
 <% $environment.each |$key, $entry| { -%>
 Environment="<%= $key %>=<%= $entry['value'] %>"
 <% } -%>
+<% unless $dropin_config =~ Undef { -%>
+<% $dropin_config.each |$key, $value| { -%>
+<%= $key %>=<%= $value %>
+<% } -%>
+<% } -%>
```
Comment (ekohl) [inline templates/jenkins-override.epp:12]: Then the `Undef` line becomes redundant
```suggestion
<% $dropin_config.each |$key, $value| { -%>
<%= $key %>=<%= $value %>
<% } -%>
```

File: manifests/init.pp [hunk 5]
```diff
@@ -281,6 +284,7 @@
   Boolean $manage_service                         = true,
   Boolean $service_enable                         = true,
   Enum['running', 'stopped'] $service_ensure      = 'running',
+  Hash $service_override                          = {},
```
Comment (ekohl) [inline manifests/init.pp:287]: If the EPP is `Hash[String, String]`, I'd make it the same here. I'd also consider `String[1]` to only allow non-empty strings, at least for the keys.

**Review by ekohl — CHANGES_REQUESTED** [review]
It would be nice to have a test for this. Either a unit test or an acceptance test (note that's an `or`, not an `xor`).

File: templates/jenkins-override.epp (hunk 4 above)
Comment (jan-win1993) [reply templates/jenkins-override.epp:12]: Sorry, I am kind of a Puppet newbie. Thanks for the recommendations.
I will take this approach.

Comment (ekohl) [reply templates/jenkins-override.epp:12]: No worries, we all had to start somewhere :)

**Review by ekohl — APPROVED** [review]
The code looks good to me. I just wonder if the test failures are related or my fixes were insufficient. I've opened https://github.com/voxpupuli/puppet-jenkins/pull/1045 which should give some answers.

**Pull request merged**